### PR TITLE
[SPARK-16756][SQL][WIP] Add `sql` function to LogicalPlan and `NonSQLPlan` trait

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR is a part of [SPARK-16576](https://issues.apache.org/jira/browse/SPARK-16576) that moves logicalPlan SQL generation code from SQLBuilder into logical operators.

Like `Expression`, this PR adds `sql` function for `LogicalPlan` and `NonSQLPlan` trait. The method will be `abstract` method. All logical plan should implement that or use trait `NonSQLPlan` explicitly.

``` scala
/**
 * Returns SQL representation of this plan. For the plans extending [[NonSQLPlan]],
 * this method may return an arbitrary user facing string.
 */
def sql: String
```

This PR updates testsuites including`LogicalPlanToSQLSuite` and `ExpressionToSQLSuite` in order to test new `sql` function, but does not remove `SQLBuilder` and its usage in `views.scala` of `sql/core`.
## How was this patch tested?

Pass the Jenkins tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -53,6 +53,40 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan) extend
 
   override def validConstraints: Set[Expression] =
     child.constraints.union(getAliasedConstraints(projectList))
+
+  override def sql: String = {
+    if (projectList.exists(expr => expr.find(e => e.isInstanceOf[NonSQLExpression]).isDefined)) {
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: This `if` is overly complex and forces a reader to read the expression inside-out (not left to right).

I can't propose anything better than beg for a boolean val with the entire boolean condition on a separate line.

WDYT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -53,6 +53,40 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan) extend
...
+      throw new UnsupportedOperationException("NonSQLExpression")
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Would `assert` be applicable here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -99,6 +133,12 @@ case class Generate(
 
     if (join) child.output ++ qualified else qualified
   }
+
+  override def sql: String = {
+    val columnAliases = generatorOutput.map(_.sql).mkString(", ")
+    s"${child.sql} LATERAL VIEW ${if (outer) "OUTER" else ""} " +
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Could you move `${if...` outside the interpolated string?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 4]
```diff
@@ -112,6 +152,11 @@ case class Filter(condition: Expression, child: LogicalPlan)
       .filterNot(SubqueryExpression.hasCorrelatedSubquery)
     child.constraints.union(predicates.toSet)
   }
+
+  override def sql: String = child match {
+    case _: Aggregate => s"${child.sql} HAVING ${condition.sql}"
+    case _ => s"${child.sql} WHERE ${condition.sql}"
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: The only difference is `HAVING` vs `WHERE`, right?

Mind extracting the small difference out and doing the following instead:

```
val havingOrWhere = ???
s"${child.sql} $havingOrWhere ${condition.sql}"
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -167,6 +212,8 @@ case class Intersect(left: LogicalPlan, right: LogicalPlan) extends SetOperation
 
     Statistics(sizeInBytes = sizeInBytes, isBroadcastable = isBroadcastable)
   }
+
+  override def sql: String = s"(${left.sql}) INTERSECT (${right.sql})"
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I think `INTERSECT` et al should be a value to later limit typos (I think there are tests that test that the output contains `INTERSET`, right?)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 6]
```diff
@@ -495,6 +573,92 @@ case class Aggregate(
       super.statistics
     }
   }
+
+  private def sameOutput(output1: Seq[Attribute], output2: Seq[Attribute]): Boolean =
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Wrap it around `{` and `}`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 7]
```diff
@@ -495,6 +573,92 @@ case class Aggregate(
...
+    output1.size == output2.size &&
+      output1.zip(output2).forall(pair => pair._1.semanticEquals(pair._2))
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Think `forall { case (left, right) => left semanticEquals right }` (or with dots) could be more readable. WDYT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 8]
```diff
@@ -495,6 +573,92 @@ case class Aggregate(
...
+
+  private def isGroupingSet(e: Expand, p: Project) = {
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Ah, so you are using `{` to wrap boolean one-liners :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 9]
```diff
@@ -692,11 +864,17 @@ case class LocalLimit(limitExpr: Expression, child: LogicalPlan) extends UnaryNo
...
+
+  override def sql: String = child.sql
 }
 
 case class SubqueryAlias(alias: String, child: LogicalPlan) extends UnaryNode {
 
   override def output: Seq[Attribute] = child.output.map(_.withQualifier(Some(alias)))
+
+  override def sql: String = child match {
+    case _ => if (child.sql.equals(alias)) child.sql else s"(${child.sql}) AS $alias"
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Why do you pattern match here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 10]
```diff
@@ -731,6 +909,20 @@ case class Sample(
...
+    case SubqueryAlias(alias, _: NonSQLPlan) =>
+      val repeatable = if (withReplacement) s" REPEATABLE ($seed)" else ""
+      s"$alias TABLESAMPLE(${upperBound * 100} PERCENT)$repeatable"
+
+    case SubqueryAlias(alias, grandChild) =>
+      val repeatable = if (withReplacement) s" REPEATABLE ($seed)" else ""
+      s"${grandChild.sql} TABLESAMPLE(${upperBound * 100} PERCENT)$repeatable $alias"
+
+    case _ =>
+      val repeatable = if (withReplacement) s" REPEATABLE ($seed)" else ""
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: `repeatable` repeated three times (and don't think it's on purpose despite the name :))

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 11]
```diff
@@ -739,6 +931,15 @@ case class Sample(
 case class Distinct(child: LogicalPlan) extends UnaryNode {
   override def maxRows: Option[Long] = child.maxRows
   override def output: Seq[Attribute] = child.output
+
+  override def sql: String = child match {
+    case Union(children) =>
+      val childrenSql = children.map(c => s"(${c.sql})")
+      childrenSql.mkString(" UNION DISTINCT ")
+
+    case _: Project =>
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Replace `_` to `p` and use it instead.

**Comment by dongjoon-hyun** [comment]
Thank you for attention, @jaceklaskowski . It sounds better. I'll update like your comments.

**Comment by dongjoon-hyun** [comment]
I close this PR since the issue is closed.

**Pull request closed without merging**